Add metrics support for endpoints [DEVC-1271]

silverjam opened the pull request:
Adds support for some basic metrics to endpoints.

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018110620-19-g805a5d2f
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=760060)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018110620-19-g805a5d2f&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018110620-19-g805a5d2f&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018110620-19-g805a5d2f
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018110620-19-g805a5d2f)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018110620-19-g805a5d2f/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018110620-19-g805a5d2f

silverjam commented on package/sbp_nmea_bridge/src/sbp.c:
```diff
@@ -17,6 +17,9 @@
 #include "sbp.h"
 
+#define METRICS_NAME "sbp_nmea_bridge"
+#define SETTINGS_METRICS_NAME (METRICS_NAME "/settings")
```
👍

silverjam commented on package/cell_modem_daemon/src/cell_modem_daemon.c:
```diff
@@ -27,6 +27,7 @@
 #include "cell_modem_settings.h"
 
 #define PROGRAM_NAME "cell_modem_daemon"
+#define SETTINGS_METRICS_NAME (PROGRAM_NAME "/settings")
```
🔨

silverjam commented on package/sbp_fileio_daemon/sbp_fileio_daemon/src/main.c:
```diff
@@ -143,7 +143,7 @@ int main(int argc, char *argv[])
   }
 
-  ctx = sbp_pubsub_create(pub_endpoint, sub_endpoint);
+  ctx = sbp_pubsub_create(PROGRAM_NAME, pub_endpoint, sub_endpoint);
```
Yes, will fix

silverjam replied:
🔨

silverjam commented on package/sbp_rtcm3_bridge/src/sbp.c:
```diff
@@ -17,6 +17,9 @@
 #include "sbp.h"
 
+#define METRICS_NAME "sbp_rtcm3_bridge"
+#define SETTINGS_METRICS_NAME (METRICS_NAME "/settings")
```
🛠

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.0.0-develop-2018110620-20-g5f9b5331
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=760147)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018110620-20-g5f9b5331&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.0.0-develop-2018110620-20-g5f9b5331&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.0.0-develop-2018110620-20-g5f9b5331
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.0.0-develop-2018110620-20-g5f9b5331)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.0.0-develop-2018110620-20-g5f9b5331/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.0.0-develop-2018110620-20-g5f9b5331

silverjam commented:
TODO: need to rebase after merge of #1000

pmiettinen commented on package/endpoint_router/endpoint_router/src/endpoint_router_load.c:
```diff
@@ -267,7 +276,9 @@ static int event_port_string(yaml_parser_t *parser, void *context, consume_str_f
 
-  consume_str_fn(port, str);
+  rc = consume_str_fn(port, str);
+  if (rc != 0) return rc;
```
Just `return rc` would do it.

silverjam replied:
fixed

pmiettinen commented on package/endpoint_router/endpoint_router/src/endpoint_router_load.c:
```diff
@@ -347,6 +367,33 @@ static PROCESS_FN(port_name)
+  if (s == NULL) return false;
+
+  size_t s_size = strlen(s);
+  if (s_size == 0) return false;
```
This failure is not logged. And error string suggests it was meant to be logged.

pmiettinen commented on package/libpiksi/libpiksi/include/libpiksi/metrics.h:
```diff
@@ -56,13 +56,16 @@ typedef struct {
+typedef struct pk_metrics_value_s {
+  pk_metrics_type_t type;
+  union {
+    s32 s32;
```
Is this a good idea to have variable name shadowing the variable name?

pmiettinen replied:
Nothing obviously great comes to my mind either. Maybe `value_s32`, `value_s64` ...?

silverjam replied:
I expanded `.d` to `.data` and adding a `.as_` prefix to all the union members-- I think this makes sense since the union is like a form of casting and it makes the struct deconstructions read a lot nicer `update.value.data.as_u32` instead of `update.value.d.u32`.

pmiettinen commented on package/libpiksi/libpiksi/include/libpiksi/sbp_pubsub.h:
```diff
@@ -54,7 +54,7 @@ typedef struct sbp_pubsub_ctx_s sbp_pubsub_ctx_t;
  *                          operation failed.
  */
-sbp_pubsub_ctx_t *sbp_pubsub_create(const char *pub_ept, const char *sub_ept);
+sbp_pubsub_ctx_t *sbp_pubsub_create(const char *ident, const char *pub_ept, const char *sub_ept);
```
Function doc out of sync.

pmiettinen replied:
Multiple places.

pmiettinen commented on package/libpiksi/libpiksi/src/endpoint.c:
```diff
@@ -522,15 +531,15 @@ static int start_un_listen(const char *path, int fd)
 }
 
-static int connect_un_socket(int fd, const char *path)
+static int connect_un_socket(int fd, const char *path, bool retry_start)
```
`retry_connect` would be better?

silverjam replied:
Yes, much better, fixed.

pmiettinen commented on package/libpiksi/libpiksi/src/loop.c:
```diff
@@ -41,6 +41,9 @@ struct pk_loop_s {
   uv_loop_t *uv_loop;
   uv_timer_t *timeout_timer;
   char uv_error_msg[MSG_BUF_SIZE];
+#ifdef UNIT_TEST_WORKAROUND
```
Seems shady, maybe at least some commenting?

silverjam commented:
@pmiettinen @benjaminaltieri rebased against master to bring in changes from #1000, addressed review comments, need to do a quick sanity check to make sure everything still works

silverjam commented:
sanity/smoke-test looked ok

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.1.0-develop-2018111002-22-g64f9d00b
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=764780)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.1.0-develop-2018111002-22-g64f9d00b&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.1.0-develop-2018111002-22-g64f9d00b&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.1.0-develop-2018111002-22-g64f9d00b
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2018111002-22-g64f9d00b)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2018111002-22-g64f9d00b/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2018111002-22-g64f9d00b

benjaminaltieri commented on package/standalone_file_logger/standalone_file_logger/src/standalone_file_logger.cc:
```diff
@@ -351,7 +355,7 @@ int main(int argc, char *argv[])
 
   /* Set up settings */
-  settings_ctx_t *settings_ctx = settings_create();
+  settings_ctx_t *settings_ctx = settings_create("standalone_file_logger/settings");
```
`settings/standalong_file_logger`?

silverjam replied:
fixed

silverjam commented:
Will wait for @pmiettinen seal of approval

anonymous-travis-swiftnav commented:
## HITL smoke tests: v2.1.0-develop-2018111002-26-g5748d439
These test runs are kicked off whenever you push a new commit to this PR. All passfail metrics in these runs must pass for the `hitl/pass-fail` status to be marked successful.
### job status
+ [live-roof-650-townsend-post runs](https://gnss-analysis.swiftnav.com/jobs/id=766115)
### gnss-analysis results
At least one run must complete for these links to have data.
#### passfail
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=pass_fail&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.1.0-develop-2018111002-26-g5748d439&groupby_key=firmware&display_type=table)
#### detailed
+ [live-roof-650-townsend-post](https://gnss-analysis.swiftnav.com/summary_type=q50&metrics_preset=detailed&scenario=live-roof-650-townsend-post&build_type=buildroot_pr&firmware_versions=v2.1.0-develop-2018111002-26-g5748d439&groupby_key=firmware&display_type=table)

anonymous-travis-swiftnav commented:
## v2.1.0-develop-2018111002-26-g5748d439
+ [Artifacts (HITL Dashboard Page)](https://gnss-analysis.swiftnav.com/artifacts/build_type=buildroot_pr&name_filter=v2.1.0-develop-2018111002-26-g5748d439)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v2.1.0-develop-2018111002-26-g5748d439/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v2.1.0-develop-2018111002-26-g5748d439

silverjam commented:
Clang format passed, merging
